src/apis/impl/AuthenticationApiImpl.ts

```diff
@@ -100,10 +102,69 @@ export class AuthenticationApiImpl implements AuthenticationApi {
     return null
   }
 
-  async initApiAndUserAuthenticationToken(process: AuthenticationProcess, validationCode: string, tokenAndKeyPairProvider: (groupId: string, userId: string) => ([string, [string, string]] | undefined)): Promise<[MedTechApi, ApiInitialisationResult]> {
+  async authenticateAndAskForAccess(userLogin: string, shortLivedToken: string, dataOwnerKeyPair: [string, string] | undefined, tokenAndKeyPairProvider: (groupId: string, userId: string) => ([string, [string, string]] | undefined)): Promise<AuthenticationResult | null> {
+    const authenticationResult = await this.initUserTokenAndCrypto(userLogin, shortLivedToken, dataOwnerKeyPair, tokenAndKeyPairProvider);
+
+    const loggedUser = await authenticationResult.medTechApi.userApi.getLoggedUser();
+    if (!loggedUser) throw new Error("User log in failed");
+    if (!loggedUser.patientId) throw new Error("User is not associated to any Patient");
+
+    const patientUser = await authenticationResult.medTechApi.patientApi.getPatient(loggedUser.patientId);
+    if (!patientUser) throw new Error(`Patient with id ${loggedUser.patientId} does not exist`);
+
+    const delegates = Object.entries(patientUser.systemMetaData?.delegations ?? {})
+      .map( (keyValue) => Array.from(keyValue[1]))
+      .flat()
+      .filter( (delegation) => !!delegation.delegatedTo)
+      .map( (delegation) => delegation.delegatedTo!);
+
+    const hcpDelegates = await authenticationResult.medTechApi.healthcareProfessionalApi.filterHealthcareProfessionalBy(
```

Comment:
You don't need the `hcpDelegates` variable : the hcp.Id is the same than the delegation.delegatedFrom. 
You can therefore directly use the list of delegatedFrom & delegatedTo in the patient. 
For Patient PATIENT_1, created by HCP_1, you take the delegation ids in bold : 
- delegatedFrom : **HCP_1** - delegatedTo : PATIENT_1;
- delegatedFrom : HCP_1 - delegatedTo : **HCP_2**;
- delegatedFrom : HCP_1 - delegatedTo : **HCP_3**;
- delegatedFrom : HCP_2 - delegatedTo : **HCP_4**;

Reply:
It could be a possibility in the future (I want to share this data with my wife/husband, ...)